zcash_primitives/src/merkle_tree.rs

```diff
@@ -225,41 +170,44 @@ impl<Node: Hashable> CommitmentTree<Node> {
 
     /// Serializes this tree as an array of bytes.
     pub fn write<W: Write>(&self, mut writer: W) -> io::Result<()> {
-        Optional::write(&mut writer, self.left, |w, n| n.write(w))?;
-        Optional::write(&mut writer, self.right, |w, n| n.write(w))?;
+        Optional::write(&mut writer, self.left.as_ref(), |w, n| n.write(w))?;
+        Optional::write(&mut writer, self.right.as_ref(), |w, n| n.write(w))?;
         Vector::write(&mut writer, &self.parents, |w, e| {
-            Optional::write(w, *e, |w, n| n.write(w))
+            Optional::write(w, e.as_ref(), |w, n| n.write(w))
         })
     }
+}
 
+impl<Node: Hashable + Clone> CommitmentTree<Node> {
     /// Adds a leaf node to the tree.
     ///
     /// Returns an error if the tree is full.
     pub fn append(&mut self, node: Node) -> Result<(), ()> {
-        self.append_inner(node, SAPLING_COMMITMENT_TREE_DEPTH)
+        self.append_inner(node, SAPLING_COMMITMENT_TREE_DEPTH_U8)
     }
 
-    fn append_inner(&mut self, node: Node, depth: usize) -> Result<(), ()> {
+    fn append_inner(&mut self, node: Node, depth: u8) -> Result<(), ()> {
         if self.is_complete(depth) {
             // Tree is full
             return Err(());
         }
 
-        match (self.left, self.right) {
+        match (&self.left, &self.right) {
             (None, _) => self.left = Some(node),
             (_, None) => self.right = Some(node),
             (Some(l), Some(r)) => {
-                let mut combined = Node::combine(0, &l, &r);
+                let mut combined = Node::combine(0.into(), l, r);
                 self.left = Some(node);
                 self.right = None;
 
                 for i in 0..depth {
-                    if i < self.parents.len() {
-                        if let Some(p) = self.parents[i] {
-                            combined = Node::combine(i + 1, &p, &combined);
-                            self.parents[i] = None;
+                    let i_usize = usize::from(i);
+                    if i_usize < self.parents.len() {
+                        if let Some(p) = &self.parents[i_usize] {
+                            combined = Node::combine((i + 1).into(), p, &combined);
```

Comment:
```suggestion
                            combined = Node::combine((i_usize + 1).into(), p, &combined);
```
The `u8` addition in the code before this suggestion happens not to be able to overflow because `i < depth` and depth is a `u8`, so `i < 255u8`, but this makes it more obvious.

Reply:
The `.into()` here is `u8 -> Level`.

Reply:
Oh this code is deleted. Should have looked at the overall diff as well.